impl import_task

mario-kr opened the pull request:
Hey, i just implemented it for &str. it might be useful to also implement it like
```
pub fn import_tasks(r : Read) -> Result<Vec<Task>>;
```
with line splitting in the function. I think the BufReader might be required for this, so maybe instead of `r : Read` a `r : BufRead`, or maybe it is possible to construct a BufRead from a Read.

matthiasbeyer commented:
Tests are failing.

I'd say you should implement this using `BufRead`, yes.

mario-kr commented:
`cargo test` doesnt show any failing tests. travis told me it's because of an unused import (the lint unused import is being used) in annotation.rs, in which i didnt do anything

matthiasbeyer commented:
Please wait until #27 is merged and then rebase onto master.

matthiasbeyer commented on src/import.rs:
```diff
@@ -25,6 +26,16 @@ pub fn import_task(s : &str) -> Result<Task> {
 
+/// Reads line by line and tries to parse a task-object per line.
+pub fn import_tasks<BR: BufRead>(r : BR) -> Result<Vec<Task>> {
+    let vt = r.lines().filter_map(|l| l.ok())
```
Can you add a linebreak before `lines()`, so you have

```rust
let vt = r
    .lines()
    .filter_map(/* ... */)
    /* ... */
```

matthiasbeyer replied:
With the above comment, this yields outdated.

matthiasbeyer commented on src/import.rs:
```diff
@@ -25,6 +26,16 @@ pub fn import_task(s : &str) -> Result<Task> {
 }
 
+/// Reads line by line and tries to parse a task-object per line.
+pub fn import_tasks<BR: BufRead>(r : BR) -> Result<Vec<Task>> {
```
Why not changing the implementation to return a `Vec<Result<Task>>`? That would really simplify and enhance the implementation, as errors could be reported to the user of the function more easily (and not ignored as with the current impl).

matthiasbeyer commented on src/import.rs:
```diff
@@ -130,3 +141,18 @@ fn test_one_single() {
+    assert!(imported.len() == 2);
+    assert!(imported[0].status() == &TaskStatus::Waiting);
+    assert!(imported[1].status() == &TaskStatus::Waiting);
+}
```
The test really should check for all values that are passed in the JSON.

matthiasbeyer commented:
I think this makes great progress and I'm really looking forward to merge it!

mario-kr commented on src/import.rs:
```diff
@@ -17,6 +18,24 @@ pub fn import<R: Read>(r: R) -> Result<Vec<Task>> {
+        .filter(|s| s.len() > 0)
+        .map(|s| import_task(s.as_str()))
+        .filter_map(|t| t.ok())
+        .collect();
```
Basically, on errors an empty vector is returned. How do i return an Err with this iterator-way of doing things? if i make a closure, Err is returned for the closure, and not the underlying operation.

matthiasbeyer replied:
You could do it this way: 

```rust
r.lines()
    filter_map(|elem_result| {
        elem_result.and_then(|elem| {
            if elem.len() > 0 { None }
            else { import_task(s.as_str()) }
        })
    })
```

The above should be a working function body for this function. I did not test it, though.

mario-kr replied:
I dont really think this would work. as i see it, you return None if an elem (a String) has a length greater 0, and try to call import_task on an empty String. Even if i reversed the if/else blocks, It still filters out any Result::Errs without returning them to the caller, so not much different from mine. I guess the right approach might be:
```rust
for line in r.lines() {
    let mut tasks = Vec::new();
    let line = match line {
        Ok(val) => val,
        Err(e) => return TaskError::new(TaskErrorKind::ReadError, Some(Box::new(e)))
    }
    tasks.push(match import_task(line.as_str) {
        Ok(val) => val,
        Err(e) => { /* transform to TaskError and return */ }
    });
    Ok(tasks)
}
```
With this we would have the all-or-nothing i think is appropriate, and Errors arent simple swallowed by Result::ok()/Result::and_then() or similar.
If you insist on a Vec<Result<Task>>, doing it with and_then seems like the right way to me.

matthiasbeyer replied:
> I dont really think this would work. as i see it, you return None if an elem (a String) has a length greater 0, and try to call import_task on an empty String.

Yeah, I messed that up. You have to change the comparison operator. And no, the `None` Just filters out, the `Some(x)` means that `x` gets returned. So the following sentence

> Even if i reversed the if/else blocks, It still filters out any Result::Errs without returning them to the caller, so not much different from mine.

is not true.

What is not in my code example is error wrapping... that has to be done.. but then I think this should work.

---

The code example looks really good, but IMHO we should really return `Vec<Result<>>` rather than the other way round. We should not repeat the things I messed up in the interface for `import()` here! :smile:

mario-kr commented:
Hey, i changed it to `Vec<Result<Task>>` like you suggested. I also added deeper testing for import_task. import_tasks uses this function, so the Tasks we get are alright if the Result is ok.
On a side note: This will brake imag-todo, as it expects a `Result<Vec<Task>>`.

matthiasbeyer commented:
Thanks for changing. I will have a closer look today evening.

matthiasbeyer commented:
Looks good I think.

matthiasbeyer commented:
bors merge
